src/rdkafka_msg.c

```diff
@@ -330,6 +354,188 @@ int rd_kafka_msg_new (rd_kafka_topic_t *rkt, int32_t force_partition,
+                                        "VTYPE_HEADER and VTYPE_HEADERS "
+                                        "are mutually exclusive");
+                                goto err;
+                        }
+
+                        if (unlikely(!hdrs))
+                                hdrs = rd_kafka_headers_new(8);
+
+                        err = rd_kafka_header_add(hdrs,
+                                                  vu->u.header.name, -1,
+                                                  vu->u.header.val,
+                                                  vu->u.header.size);
+                        if (unlikely(err)) {
+                                error = rd_kafka_error_new(
+                                        err,
+                                        "Failed to add header: %s",
+                                        rd_kafka_err2str(err));
+                                goto err;
+                        }
+                        break;
+
+                case RD_KAFKA_VTYPE_HEADERS:
+                        if (unlikely(hdrs != NULL)) {
+                                error = rd_kafka_error_new(
+                                        RD_KAFKA_RESP_ERR__CONFLICT,
+                                        "VTYPE_HEADERS and VTYPE_HEADER "
+                                        "are mutually exclusive");
+                                goto err;
+                        }
+                        app_hdrs = vu->u.headers;
+                        break;
+
+                default:
+                        error = rd_kafka_error_new(
+                                RD_KAFKA_RESP_ERR__INVALID_ARG,
+                                "Unsupported VTYPE %d", (int)vu->vtype);
+                        goto err;
+                }
+        }
+
+        rd_assert(!error);
+
+        if (unlikely(!rkt)) {
+                error = rd_kafka_error_new(
+                        RD_KAFKA_RESP_ERR__INVALID_ARG,
+                        "Topic name or object required");
+                goto err;
+        }
+
+        rkm = rd_kafka_msg_new0(rkt,
+                                rkm->rkm_partition,
+                                rkm->rkm_flags,
+                                rkm->rkm_payload, rkm->rkm_len,
+                                rkm->rkm_key, rkm->rkm_key_len,
+                                rkm->rkm_opaque,
+                                &err, NULL,
+                                app_hdrs ? app_hdrs : hdrs,
+                                rkm->rkm_timestamp,
+                                rd_clock());
+
+        if (unlikely(err)) {
+                error = rd_kafka_error_new(
+                        err,
+                        "Failed to produce message: %s",
+                        rd_kafka_err2str(err));
+                goto err;
+        }
+
+        /* Partition the message */
+        err = rd_kafka_msg_partitioner(rkt, rkm, 1);
```

Comment:
surprised to see this here, probably some unnecessary duplication of stuff, but not a big deal.

Reply:
The produceva() code is a direct copy of producev().